File: Projects-AEB/R-AEB/cpp/src/cuboids/cuboids_api.cpp
// cuboids_raeb.cpp
#include "cuboids_api.h"
#include "cuboids_cyl3dbox_impl.hpp"

#include <memory>
#include <stdexcept>

// ------------- 全局实例定义 -------------
static std::unique_ptr<Cyl3DBoxEstimator> g_3dest_raeb;

// ------------- 初始化 -------------
void cuboids_InitRAEB(
    const nlohmann::json::object_t& whitelist,
    const nlohmann::json::object_t& cyl_cam
) {
    // 对应 Python:
    // g_3dest_raeb = Cyl3DBoxEstimator(cyl_cam=cyl_cam, whitelist=whitelist)
    g_3dest_raeb = std::make_unique<Cyl3DBoxEstimator>(
        cyl_cam,
        whitelist
    );
}

// ------------- 3D 盒计算 -------------
cv::Mat cuboids_ProcRAEB(
    const cv::Mat& objs,
    double grounding_z,
    const cv::Mat& masks
) {
    if (!g_3dest_raeb) {
        throw std::runtime_error("Cyl3DBoxEstimator 未初始化：请先调用 cuboids_InitRAEB(...)");
    }


    cv::Mat cuboids = g_3dest_raeb->cuboids_from_boxes(
        objs,           // (N,7) [x1,y1,x2,y2,conf,cls,theta_rel]
        grounding_z,    // z_world 先验 / fallback
        masks          // 3D (N,H,W) CV_8U
    );

    return cuboids;
}

// ------------- 3D 盒可视化 -------------
cv::Mat cuboids_VisCuboids(
    const cv::Mat& cyl_img_ori,
    const cv::Mat& cuboids,
    const std::vector<std::string>* labels,
    bool show_info
) {
    if (!g_3dest_raeb) {
        throw std::runtime_error("Cyl3DBoxEstimator 未初始化：请先调用 cuboids_InitRAEB(...)");
    }

    cv::Mat vis_img_3d = g_3dest_raeb->draw_3dboxes_on_cyl(
        cyl_img_ori,
        cuboids,
        1,
        labels,
        8,
        nullptr,
        show_info
    );

    return vis_img_3d;
}

File: Projects-AEB/R-AEB/python/visper/cuboids/cuboids_api.py
from __future__ import annotations

import numpy as np
from visper.cuboids.cuboids_cyl3dbox_impl import Cyl3DBoxEstimator

# 全局实例（初始化前为 None）
g_3dest_raeb: Cyl3DBoxEstimator | None = None

def cuboids_InitRAEB(
        whitelist: dict[str, dict],
        cyl_cam: dict[str, float],
) -> None:
    """
    初始化 2D 分割后处理与 3D 落点-尺寸估计器（单图版）。
    """
    global g_3dest_raeb

    g_3dest_raeb = Cyl3DBoxEstimator(
        cyl_cam=cyl_cam,
        whitelist=whitelist,
    )


def cuboids_ProcRAEB(
        objs: np.ndarray,
        grounding_z: float,
        masks: np.ndarray,
        track_ids: np.ndarray | None = None,
):

    # 3D：接地点→尺寸→盒（基于原图坐标的接地点像素进行投影）
    cuboids = g_3dest_raeb.cuboids_from_boxes(
        objs=objs,
        z=grounding_z,
        masks=masks,
        track_ids=track_ids,
    )

    return cuboids


def cuboids_VisCuboids(
        cyl_img: np.ndarray,
        cuboids: np.ndarray,
        labels: list[str] | None = None,
        show_info: bool = False,
):
    vis_img_3d = g_3dest_raeb.draw_3dboxes_on_cyl(
        cyl_img=cyl_img,
        cuboids=cuboids,
        labels=labels,
        show_info=show_info,
    )

    return vis_img_3d
